open-apparel-registry/open-apparel-registry
# Allow for reporting a facility as closed or reopened

## Overview

Adds models and an API to support marking a facility as closed or opened.

Connects #1136 

## Demo

<img width="950" alt="Screen Shot 2021-02-03 at 11 39 59 AM" src="https://user-images.githubusercontent.com/21046714/106779766-5157ff80-6615-11eb-8991-0afeb5d23cdb.png">
<img width="951" alt="Screen Shot 2021-02-03 at 11 39 36 AM" src="https://user-images.githubusercontent.com/21046714/106779689-3d140280-6615-11eb-9433-abff65736c3f.png">
<img width="948" alt="Screen Shot 2021-02-03 at 11 39 13 AM" src="https://user-images.githubusercontent.com/21046714/106779725-47ce9780-6615-11eb-909a-eba671b36de2.png">

## Testing Instructions

* Run `vagrant ssh` and `./scripts/manage migrate` and `./scripts/server`
* Go to [:6543](http://localhost:6543/) and save the OAR ID of a facility 
- Go to http://localhost:8081/api/docs/#!/facilities/facilities_read and authenticate your user.
- Fetch the facility by the ID you saved
    - The facility should have `is_closed: null`
- Go to http://localhost:8081/api/docs/#!/facilities/facilities_report
- Enter the facility id and 
    ``` 
    {
         "closure_state": "CLOSED",
         "reason_for_report": "This facility was closed."
     } 
     ```
- Submit. 
    - The request should return a FacilityActivityReport with `closure_state: ‘CLOSED’` and `status: ‘PENDING’` and appropriate user, contributor, and created/updated date times. Note the ID.
- Go to http://localhost:8081/api/docs/#!/facility-activity-reports/facility_activity_reports_approve and enter the report ID and ```{
    "status_change_reason": "The facility report was confirmed."
}```
- Submit. 
    - The request should return the same FacilityActivityReport with `status: “CONFIRMED”`, all of the status_change_ fields filled out, and `approved_at` set to the current date time.
- Go to http://localhost:8081/api/docs/#!/facilities/facilities_read and fetch the facility again
    - It should have `is_closed: true`
- Repeat the process with a new report, but enter `"closure_state": “OPEN”`
    - Approve the facility report. The facility should now have `is_closed: false`
- Create a new facility report with `"closure_state": “CLOSED”`
- Go to http://localhost:8081/api/docs/#!/facility-activity-reports/facility_activity_reports_reject 
- Enter the facility ID and ```{
    "status_change_reason": "The facility report is incorrect."
}```
    - Submit. The response should have `"status": "REJECTED”` and the status_change_ fields should be filled out appropriately. 
- Fetch the facility. 
     - Its status should not have changed. 

## Checklist
